### ISSUE TITLE
Some thoughts on enhancing OC's permissions

### ISSUE THREAD

**Issue by bakintunde** [OrchardCMS/OrchardCore#4040]
A couple of thoughts on enhancements around content permissions for OC

Permissions could become primarily, a function of three things: 

1. **Which specific content items a user/subject has access to**: OC does not have this at the moment. This can be driven using a new construct called **_groups_** that can be defined in a hierarchical manner to denote team structures

2. **What they can do with the content items they have access to:** This is currently achieved using _roles_ and _permissions_
3. The **worklows** defined on the content type and at which points in the workflow the user can be involved.

To achieve this with OC, we might do the following:

- Enhance ContentType Definitions to include **Content Item Access Rules**. This could be used to filter views of data based on who's logged in. These rules would map a set of content items to a set of users and might contain fields like:
  - The set of content items defined with for example an SQL-like clause e.g. 'Location = 'Redmond' and Classification = 'Secret' that limits the content item records affected by the rule
  - The set of users may be collectively defined by one or more of the following:
     - One or more groups of users. Specifying a top level group implies users in child groups are affected
     - Roles : one or more roles
     - One or more users 
     - An expression that could be based on a user's profile properties.
  - A toggle indicating whether the access rule is permissive or prohibitive 
- Create a built-in **UserProfileContentPart**  that allows the user's profile to be extended and used in expressions in the Content Item Access Rules 
- Enhance **ContentTypeAuthorizationHandler** to limit access using Content Item Access Rules and limit operations using assigned role permissions
- Enhance Controllers in **OrchardCore.Contents** module to filter records by using Content Item Access Rules
- Content item remains owned by the user that creates it. 
- Implement immutable id on User that is referenced by permission settings, group, content access rules etc.
- define a **HumanWorkflowActionState** content part for content types that will participate in a "human" workflow leveraging the regular workflow. It could contain these fields.
  - Ordered list of possible workflow states for the content item (e.g. Draft, Initial, Final) - different from CMS Published or Unpublished state. This means a content item could go through these workflow states behind the scenes (Unpublished) or in public (published)
   - List of possible UI action buttons which are typically Approve, Reject that should be available when taking the content item through a workflow instance.  Where these UI action buttons lead will depend on what's configured in the workflow for the content type
   - Mapping of Human workflow states to UI Actions to indicate which are available at different states. E.g.:
      - At Draft state, Approve UI action moves to Review state, (actual path to Review state is defined in workflow. Multiple steps may exist on that path) 
      - At Review - (Reject moves to Draft or Approve to Final)
      - At Final - only reject is available (to a select group of users) which moves state back to Draft 
   - Workflow Type Id - a reference to the specific workflow type the content item is part of
  - Workflow Instance Id - contains a reference to the actual instance of the running workflow, the content item is currently involved in. Might be useful if workflow instance is deleted on completion.
  - A dynamic collection of data items generated/used along the workflow e.g. a unique url to take the user after the workflow
- We further want to limit access to the UI action buttons with **UI Action Access Rules** which may be defined by: 
   - One more script expressions based on the user's profile or objects in the OC environment e.g. check for elapsed time since creation, a user's profile property etc
   - One or more groups
   - One or more roles
   - One or more user Ids
   - One or more permissions.
- UI Action Access Rules and Content Item Access Rules could leverage a reusable set of rules from a new "Rules Module"
- Create a Human Workflow Actions Workflow task that works a bit like the Script task but instead of a script, the outcome is to be set by a user's click of available UI Actions defined in the WorkflowActionState content part on the content type e.g. Approve, Reject, Hold, etc

**Comment by bakintunde** [comment]
cc 
@Skrypt 
@jean 
@deanmarcussen 
@petedavis 
@sebastienros

**Comment by jean** [comment]
Some feedback ..

In the following I'm referencing DCWorkflow, an implementation that has served Plone well for about 20 years.
Some links: 

- Good high-level description: [DCWorkflow's hidden gems](http://www.martinaspeli.net/articles/dcworkflows-hidden-gems)
- Reference docs: [Workflow](https://docs.plone.org/external/plone.app.dexterity/docs/advanced/workflow.html)
- Training material: [Plone Workflow](https://training.plone.org/5/workflow/index.html)

> Permissions could become primarily, a function of three things:

Permissions are assigned to roles. 
Roles are evaluated with respect to Users.
A user can have roles implicitly (globally, e.g. Authenticated, Anonymous; or locally, e.g. Owner), explicitly (e.g. Admin), or due to group membership (the Editors group has the Editor role).
For a blog, users could belong to the Authors group, which has the Author role, which has the "Create blog posts" permission.
We could say that in our case editors are also authors, so the Editors group contains the Authors group, and editor users get the "Create blog posts" permission from there.

> - Which specific content items a user/subject has access to: OC does not have this at the moment. This can be driven using a new construct called groups that can be defined in a hierarchical manner to denote team structures

Yes.

- A group contains users and groups.
- A group has a set of roles, globally.
- A group can be assigned *local* roles on content items. 
  - Content items might also be selected by "a SQL-like clause e.g. 'Location = 'Redmond' and Classification = 'Secret'" on the group.
  - It would be nice if contained items (in a Bag) could inherit the group setting of the container.

> - What they can do with the content items they have access to: This is currently achieved using roles and permissions
> 
> - The workflows defined on the content type and at which points in the workflow the user can be involved.

We should be careful of the definition of "workflow" that we're using. 
Orchard already has powerful workflows that have nothing to do with specific content types, permissions or item states.
Are we proposing an extension to this, or are we talking about something different?

> To achieve this with OC, we might do the following:
> 
> - Enhance ContentType Definitions to include Content Item Access Rules. This could be used to filter views of data based on who's logged in. These rules would map a set of content items to a set of users and might contain fields like:

I think this is wrong. Content Types can be used in different contexts that require different access patterns.
They can partake in different workflows, sequentially or even simultaneously (for example, an object may be published, but require translation. You can track the review state in the main workflow and the translation state in another).

The user/group/role/permission machinery should work independently from content type definition.

>   - The set of content items defined with for example an SQL-like clause e.g. 'Location = 'Redmond' and Classification = 'Secret' that limits the content item records affected by the rule
>   - The set of users may be collectively defined by one or more of the following:
>     - One or more groups of users. Specifying a top level group implies users in child groups are affected
>     - Roles : one or more roles
>     - One or more users
>     - An expression that could be based on a user's profile properties.
>   - A toggle indicating whether the access rule is permissive or prohibitive

I'd rather just add users to groups manually or via a workflow that picks the right group.

> - Create a built-in UserProfileContentPart that allows the user's profile to be extended and used in expressions in the Content Item Access Rules
> - Enhance ContentTypeAuthorizationHandler to limit access using Content Item Access Rules and limit operations using assigned role permissions

Access should be evaluated based on the permissions a user has in some context due to their global and local roles, including group roles.
I don't think this requires a concept of "Content Item Access Rules". Just evaluate access based on permissions.

> - Enhance Controllers in OrchardCore.Contents module to filter records by using Content Item Access Rules
> - Content item remains owned by the user that creates it.
> - Implement immutable id on User that is referenced by permission settings, group, content access rules etc.
> - define a HumanWorkflowActionState content part for content types that will participate in a "human" workflow leveraging the regular workflow. It could contain these fields.
>   - Ordered list of possible workflow states for the content item (e.g. Draft, Initial, Final) - different from CMS Published or Unpublished state. This means a content item could go through these workflow states behind the scenes (Unpublished) or in public (published

I think there should be as few concepts as possible. 
There should be no difference between CMS Published/Unpublished state and "workflow-defined" states. 
If the Anonymous role has View permission on a state, it is de facto published.

The states aren't ordered: instead, each state defines transitions to target states. 
This implies a graph of states.

>   - List of possible UI action buttons which are typically Approve, Reject that should be available when taking the content item through a workflow instance. Where these UI action buttons lead will depend on what's configured in the workflow for the content type

A state defines transitions that are guarded by permission or scripts, and that can be rendered as e.g. buttons or links if guard conditions indicate that the user has the right to actuate the transition, or can be triggered automatically if the guard conditions are true.

Each item will need to have a role-permission map that is updated upon each state transition. 

E.g. on a Blog post, the mapping is updated as the post moves through workflow states:

- Draft
  - Owner: Can view, Can edit, Can submit for review (-> Pending review)
  - Editor: no permission
  - Anonymous: no permission
- Pending review
  - Owner: Can view, Can withdraw (-> Draft)
  - Editor: Can view, Can publish (-> Published), Can submit for revision (-> Draft), Can archive (-> Archived)
  - Anonymous: no permission
- Published
  - Owner: Can view
  - Editor: Can view, Can submit for revision (-> Draft), Can archive (-> Archived)
  - Anonymous: Can view
- Archived
  - Owner: Can view
  - Editor: Can view, Can submit for revision (-> Draft)
  - Anonymous: no permission

**Comment by bakintunde** [comment]
Thanks for the feedback.

I realize there's a lot of stuff here. I will try to unpack this into the different user stories in view later but I'll put my responses inline here for now. These ideas are still preliminary.

The proposed "groups" and "content item access rules" are to extend OC's permissions capability to achieve limiting a set of users to sets of content items unlike now where users with a view permission on a content type can view any of its items. Permissions will continue being defined on roles (and only roles) i.e. "what can be done with what is seen/accessed".

>Some feedback ..
>
>In the following I'm referencing DCWorkflow, an implementation that has served Plone well for about 20 years. (It's probably been the most stable feature of the CMS.) Whatever we do, it should be based on proven prior work. If there are other good examples to study, please link 'em.
>Some links about DCWorkflow:
> * Good high-level description: DCWorkflow's hidden gems
> * Reference docs: Workflow
> * Training material: Plone Workflow

A similar separation of roles and groups can be seen in Salesforce(SF). The term "role" in SF is analogous to what I'm proposing as groups for OC, the thinking is essentially the same - a hierarchical structure that determines which content items you have access to.

See - Difference between a profile (like OC Roles) and a role (proposed OC Group) in Salesforce
https://success.salesforce.com/answers?id=90630000000gnsLAAQ

No need to reinvent any wheels but leverage what exists in OC.

>Permissions are assigned to roles.
Roles are evaluated with respect to Users.
A user can have roles implicitly (globally, e.g. Authenticated, Anonymous; or locally, e.g. Owner), explicitly (e.g. Admin), or due to group membership (the Editors group has the Editor role).
For a blog, users could belong to the Authors group, which has the Author role, which has the Create blog posts permission.
We could say that in our case editors are also authors, so the Authors group is a member of the Editors group, and editor users get the Create blog posts permission from there.

We wouldn't want to deviate from the existing metaphors in OC without strong reason. All roles are essentially the same in OC except for Anonymous and Authenticated which are treated specially internally. You can still add these to the UI and configure custom roles for them. Having to add users to "groups" which have "roles" is not needed in OC as you can already "imply" multiple permissions in one by passing the name of an over-arching permission in the ImpliedBy parameter of the [Permission class constructor](https://github.com/OrchardCMS/OrchardCore/blob/67e0aa21f62f6477ed3de79a4c436aec3b94d9bc/src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Security/Permissions/Permission.cs#L26). See an example in the [OC training demo module](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/51772eea0024fe18cc11ed7d9513701f118a0bd1/Permissions/PersonPermissions.cs#L17). Note that the containing relationship is not currently obvious in the UI.
If a specific permission is needed by different roles, simply explicitly assign it to both.
So let's have users put in groups ONLY to reflect team structure and by implication, selective access to content which is more intuitive in my opinion.

>Yes.
>
>A group contains users and groups.
A group has a set of roles, globally.
A group can be assigned local roles on content items.
Content items might also be selected by "a SQL-like clause e.g. 'Location = 'Redmond' and Classification = 'Secret'" on the group.
It would be nice if contained items (in a Bag) could inherit the group setting of the container.

I don't see a need to directly associate groups with a specific roles. In reality, groups/teams typically have members with different roles. Also a logical distinction of roles as local/global doesn't do anything for me in the context of OC. Groups will simply contain other groups or users similar to organizational structure as you might have in LDAP forests.

>We should be careful of the definition of "workflow" that we're using.
Orchard already has powerful workflows that have nothing to do with specific content types, permissions or item states.
Are we proposing an extension to this, or are we talking about something different?

The proposal is to leverage the existing workflow system to achieve human workflow. Designated content types when involved in a workflow will have states/human workflow steps which will provide a set of actions users can perform through the UI(think buttons to click like 'Approve/Reject'). Use of the buttons would be permissioned.

>I think this is wrong. Content Types can be used in different contexts that require different access patterns.
They can partake in different workflows, sequentially or even simultaneously (for example, an object may be published, but require translation. You can track the review state in the main workflow and the translation state in another).

>The user/group/role/permission machinery should work independently from content type definition.

It's not wrong. It is because of the different contexts that we should have multiple access rules on a content type. Essentially, each instance of a rule on a content type could say:
For a specific class of user (role,group,user) under certain scenarios (defined by an expression describing a set of conditions e.g. time of day, workflow state) show/hide a certain set of content items (defined by a filter) for a content type. Rules would be evaluated from top to bottom - whichever rule evaluates to true first will be used.
Content item access rules isn't about workflows but which records a user can see.

In my opinion, concurrent workflows on the same record is probably overkill at this point. To use your example, if an object is published and needs translation, the translation will go through it's own workflow. Even if concurrent workflows are needed in future, it might  be achieved with perhaps multiple instances of the Human Workflow enabling contentPart on the content item.

User/group/role/permission does not work independently of content type definitions even now. Once a content type is created, OC implicitly provides built-in CRUD type permissions for it that can be assigned to roles. The proposal is to somehow extend that permissions system so we can limit certain permissions to certain users.

>I'd rather just add users to groups manually or via a registration workflow (in the current Orchard sense) that picks the right group.

Yes, users would be assigned to groups. A group would contain users or groups. The registration workflow you linked to is for legacy Orchard, not Orchard Core (OC). The ability to extend the user profile is not yet in OC. There is standard registration workflow but the workflow can be configured to achieve it.

>Access should be evaluated based on the permissions a user has in some context due to their global and local roles, including group roles.
I don't think this requires a concept of "Content Item Access Rules". Just evaluate access based on permissions.

See first point above. For example, we'd like to be able to fulfill the requirement - "show users in fulfillment Group only orders that have a _Paid_ state"

>I think there should be as few concepts as possible.
There should be no difference between CMS Published/Unpublished state and "workflow-defined" states.
If the Anonymous role has View permission on a state, it is de facto published.
>
>The states aren't ordered: instead, each state defines transitions to target states.
This implies a graph of states. This short video of a workflowmanager for Plone illustrates this.

I think we should rather ensure every concept introduced is justified. I find that hacking things in unintuitive ways in an attempt to _minimize_ complexity is actually counter-productive and creates technical debt. In reality, content items do go through ordered workflows with permissioned access. A record may become finalized and only a few users ever get to interact with it.
Salesforce has this concept for [sales opportunities](https://help.salesforce.com/articleView?id=000338431&type=1&mode=1).

The idea is to define mappings of user actions at different workflow states to transitions in the HumanWorkFlowActionState(should probably change this name) contentPart. This mapping will be used when the workflow hits a (proposed)human workflow step.

>A state defines transitions that are guarded by permission or scripts, and that can be rendered as e.g. buttons or links if guard conditions indicate that the user has the right to actuate the transition, or can be triggered automatically if the guard conditions are true.

>Each item will need to have a role-permission map that is updated upon each state transition.

>E.g. on a Blog post, the mapping is updated as the post moves through workflow states:

>Draft
Owner: Can view, Can edit, Can submit for review (-> Pending review)
Editor: no permission
Anonymous: no permission
Pending review
Owner: Can view, Can withdraw (-> Draft)
Editor: Can view, Can publish (-> Published), Can submit for revision (-> Draft), Can archive (-> Archived)
Anonymous: no permission
Published
Owner: Can view
Editor: Can view, Can submit for revision (-> Draft), Can archive (-> Archived)
Anonymous: Can view
Archived
Owner: Can view
Editor: Can view, Can submit for revision (-> Draft)
Anonymous: no permission

This example workflow is what we're trying to support. I don't think a role-permisison map needs to be updated on each transition for each workflow instance. The rules could be defined once and evaluated at every step.

I lean towards introducing these changes in one or more separate modules for now so they can be optionally turned on and minimize conflict with the core of OC.

**Comment by jean** [comment]
TL;DR: I believe that setting permissions once globally is a major limitation of Orchard, DCWorkflow provides an example of setting permissions per state and per item, and I'm trying to explain what DCWorkflow does.

----

OK, I'm going to use item-state for the type of state I'm talking about.
In DCWorkflow this is the only type of state that exists, but Orchard we also have to consider activity-based workflow states that are independent of content.

A DCWorkflow item-state graph is something totally different from an Orchard workflow, which is basically a visual program that hooks into CMS events. 
In the Plone context, there are [Content Rules](https://docs.plone.org/working-with-content/managing-content/contentrules.html) which address the use cases that Orchard workflows address.
(I think Orchard workflows are better at this than Content Rules, and in fact once upon a time Plone also had an activity-based workflow implementation called [AlphaFlow](https://web.archive.org/web/20060313134707/http://www.gocept.com/gocept/activities/innovation/alphaflow?set_language=en&cl=en), but the community opted to stick with the simpler DCWorkflow and much later introduced Content Rules.)

The DCWorkflow states that content items transition through is completely separate from the line-of-business processes that are automated via content rules (Plone) or workflows (Orchard).
States in DCWorkflow are only about managing the permissions per state, and the transitions between states.
There are touching points, e.g. a content rule can trigger an item-state transition.

> The proposed "groups" and "content item access rules" are to extend OC's
> permissions capability to achieve limiting a set of users to sets of content
> items unlike now where users with a view permission on a content type can
> view any of its items. Permissions will continue being defined on roles (and
> only roles) i.e. "what can be done with what is seen/accessed".

Permissions on roles, good.

Content item access rule, I'm not quite clear on that. Here's how I would understand it:
A user already has a set of permissions based on their roles.
(And, in DCWorkflow, the set of permissions of a role depends on the item-state).

A content item access rule could additionally check business rules, e.g. "a user can't create more than 1 post per hour",
so on a blog the availability of a "create post" button may be guarded by a rule.
(Such a rule would fall outside of what DCWorkflow covers, because it does not pertain to an item-state change.)
However in OC I would expect this to be a widget or layer visibility rule, not a rule defined on the content type.

In DCWorkflow, there are guards (permission checks, or expression) on transitions.
E.g. there may be an "approve" transition, guarded by the "Can Review" permission, from "Pending review" to "Pending review" (i.e. no change in item-state) that increments an "approvals" workflow variable.
And a "publish" transition from "Pending review" to "Published" that is guarded by an expression that checks that the "approvals" workflow variable is greater than 2.

> See - Difference between a profile (like OC Roles) and a role (proposed OC Group) in Salesforce

Hmm I can't really interpret this without a proper understanding of workflow in Salesforce, and I think that is such a massive system with a lot of history that we'll struggle to relate it to OC without boiling it down to something very short and sweet.

> No need to reinvent any wheels but leverage what exists in OC.

Yeah of course. What are you specifically thinking about regarding reinventing wheels?

> We wouldn't want to deviate from the existing metaphors in OC without strong reason.

Where is this deviating from OC metaphors?
Currently, roles are already assigned to users. We're discussing introducing the concept of Groups so that we can assign roles to groups of users. 
We want to assign a role to a group of users because we want to give some permissions to a group of users (and the role has the permissions), e.g. we want company representatives to be able to update the company profile. 

> Having to add users to "groups" which have "roles" is not needed in OC as you can already "imply" multiple permissions in one by passing the name of an over-arching permission in the ImpliedBy parameter of the [Permission class constructor](https://github.com/OrchardCMS/OrchardCore/blob/67e0aa21f62f6477ed3de79a4c436aec3b94d9bc/src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Security/Permissions/Permission.cs#L26).

Hierarchical permissions are indeed cool. 
However I think it is orthogonal to the use of groups, because we don't want to assign (an over-arching permission) to users individually, we want to assign it to groups of users. 

> See an example in the [OC training demo module](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/51772eea0024fe18cc11ed7d9513701f118a0bd1/Permissions/PersonPermissions.cs#L17). Note that the containing relationship is not currently obvious in the UI.

"everybody who has the ManagePersons permission" <-- in DCWorkflow, this would be _everybody who has the Manager role_.
The Manager role has the ManagePersons permission.
And the quick way to assign the Manager role to many people is to create a Managers group, add the Manager role to the group, and add all the relevant users to the group.

> If a specific permission is needed by different roles, simply explicitly assign it to both.

Yes, that doesn't change.
E.g. the Editor role has View and Edit permission (in some item-states), and the Author role also has View and Edit permissions (in some item-states).

> So let's have users put in groups ONLY to reflect team structure and by implication, selective access to content which is more intuitive in my opinion.

Access to content (or other functionality) is determined by permission. 
You get access via group membership because a group has roles, and roles have permissions.

The result is the same, but there is only one mechanism, namely permission.

I should emphasize that the groups I'm talking about exist purely to help manage permissions.

If you want to model team or organisational structure in the CMS, you would define content types such as Department, Committee, Faculty, PostGrads, etc., if you were dealing with a university.
Or in another case define it via taxonomy.

> I don't see a need to directly associate groups with a specific roles. 

The requirement is to assign permissions to groups of users. 
A user has a permission because they have a role that grants the permission.
You either need to assign the role to each user, or assign the role to a group, and put all those users in the group.
This meaningfully labels that set of users (Authors, Editors, Students, Members, etc.).
Depending on the use case, you might create new groups programmatically, e.g. create a "Staff of {{CompanyName}}" group when the company item is created.

I suspect you imply writing an access rule, which grants permission to the set of users matched by the rule.
Maybe your "access rule" corresponds roughly to my "group"?

> In reality, groups/teams typically have members with different roles.
> Also a logical distinction of roles as local/global doesn't do anything for me in the context of OC.

"Staff of {{CompanyName}}" would have the **local** role of Author on TheCompanyItem (and items in its bag?).
If you want to be able to give users different access in different contexts, you need the concept of local roles in some sense.

In DCWorkflow, this is very item-based. In Orchard this would probably be more natural to define via expressions on the group.
E.g. in the above case the scope of "Staff of {{CompanyName}}" could be defined like 
"Editor": `Model.ContentItem.ContentItemId == COMPANYID` (_users in the "Staff of {{CompanyName}}" group get the Editor role on the Company content item_).

The scope of an "Editors" group could be "Editor": `Model.ContentItem.ContentType == "BlogPost"` (_users in the "Editors" group get the Editor role on all blog posts_).

(In the item-based way of working, one would give the Editors group the local role Editor on a Blog item,
all BlogPost items would be contained within the parent Blog item,
and so members of the group would inherit the Editor role on all the child objects.
However I believe this hierarchical way of thinking is foreign to Orchard.)

> > We should be careful of the definition of "workflow" that we're using.
> > Orchard already has powerful workflows that have nothing to do with specific content types, permissions or item states.
> > Are we proposing an extension to this, or are we talking about something different?
> 
> The proposal is to leverage the existing workflow system to achieve human workflow. Designated content types when involved in a workflow will have states/human workflow steps which will provide a set of actions users can perform through the UI(think buttons to click like 'Approve/Reject'). Use of the buttons would be permissioned.

In this discussion there are two types of "state" in play.
Current Orchard workflow state, i.e. where in the workflow flowchart are we.
And item state, in the DCWorkflow sense, where a state denotes a matrix of permissions per role.

This departs from Orchard roles, that are declared globally.
Some role has some set of permissions. That's defined once, globally.
In a state-based workflow, a role has some set of permission *in some workflow state*. 
I.e. the Author role has Edit permission if an item is in the "Draft" state, but not in the "Pending review" state or the "Published" state.

I believe that setting permissions once globally is a major limitation of Orchard, and this is why I'm trying to describe what's nice about DCWorkflow's item-states.

> > I think this is wrong. Content Types can be used in different contexts that require different access patterns.
> > They can partake in different workflows, sequentially or even simultaneously (for example, an object may be published, but require translation. You can track the review state in the main workflow and the translation state in another).
> 
> > The user/group/role/permission machinery should work independently from content type definition.
> 
> It's not wrong. It is because of the different contexts that we should have multiple access rules on a content type. Essentially, each instance of a rule on a content type could say:
> For a specific class of user (role,group,user) under certain scenarios (defined by an expression describing a set of conditions e.g. time of day, workflow state) show/hide a certain set of content items (defined by a filter) for a content type. Rules would be evaluated from top to bottom - whichever rule evaluates to true first will be used.
> Content item access rules isn't about workflows but which records a user can see.

We're still talking past each other. 
Remember I'm describing how DCWorkflow works.
I'm not trying to figure out how a workflow system could work, I'm describing how this system has been working for a couple of decades.
And trying to show what it would look like in Orchard.

It certainly doesn't make sense to adopt wholesale, but I would like to get across what it is that DCWorkflow does.

> In my opinion, concurrent workflows on the same record is probably overkill at this point. To use your example, if an object is published and needs translation, the translation will go through it's own workflow. Even if concurrent workflows are needed in future, it might be achieved with perhaps multiple instances of the Human Workflow enabling contentPart on the content item.

I mention concurrent workflows mainly because this is just an emerging property of DCWorkflow.
I.e. you can assign more than one workflow to the same content type and it just works. 

> User/group/role/permission does not work independently of content type definitions even now. Once a content type is created, OC implicitly provides built-in CRUD type permissions for it that can be assigned to roles. The proposal is to somehow extend that permissions system so we can limit certain permissions to certain users.

The fact that there is a "Can add Article" permission doesn't imply that the role-permission mapping for that needs to be done as part of the Article content type definition.
On a personal blog you might assign a "TwoStateWorkflow" (Draft-publish->Published), 
on a news site you might assign an "EditorialWorkflow" (Draft-submit->Pending-approve->Published,Pending-reject->Draft,Pending-archive->Archived,Published-archive->Archived).

> > I'd rather just add users to groups manually or via a registration workflow (in the current Orchard sense) that picks the right group.
> 
> Yes, users would be assigned to groups. A group would contain users or groups. The registration workflow you linked to is for legacy Orchard, not Orchard Core (OC). The ability to extend the user profile is not yet in OC. There is standard registration workflow but the workflow can be configured to achieve it.

I know the registration workflow I linked to is for legacy Orchard, I just meant to give an example of automatically adding users to groups upon registration, to make it clear what I mean by it.

> > Access should be evaluated based on the permissions a user has in some context due to their global and local roles, including group roles.
> > I don't think this requires a concept of "Content Item Access Rules". Just evaluate access based on permissions.
> 
> See first point above. For example, we'd like to be able to fulfill the requirement - "show users in fulfillment Group only orders that have a _Paid_ state"

In the "Paid" item-state, the "Fulfillment Manager" role has the "View" permission.
The "Fulfillment Managers" group has the "Fulfillment Manager" role.
MrX, MrY and MsZ are in the "Fulfillment Managers" group, so they have the "View" permission.

> > I think there should be as few concepts as possible.
> > There should be no difference between CMS Published/Unpublished state and "workflow-defined" states.
> > If the Anonymous role has View permission on a state, it is de facto published.
> > The states aren't ordered: instead, each state defines transitions to target states.
> > This implies a graph of states. This short video of a workflowmanager for Plone illustrates this.
> 
> I think we should rather ensure every concept introduced is justified. I find that hacking things in unintuitive ways in an attempt to _minimize_ complexity is actually counter-productive and creates technical debt.

Yes, bad things are bad. 
What specifically are you referring to?

I think you may be objecting to what you see as a conflation of what you're calling "human" states and Orchard workflow states.
I hope I've clarified that by distinguishing between item-states and Orchard workflow states.
I don't think either type of state is specifically "human".

> In reality, content items do go through ordered workflows with permissioned access. A record may become finalized and only a few users ever get to interact with it.

Sure, I'm describing how DCWorkflow handles content items that go through ordered workflows with permissioned access.
Transitions don't necessarily need to be triggered manually.

> Salesforce has this concept for [sales opportunities](https://help.salesforce.com/articleView?id=000338431&type=1&mode=1).
>
> The idea is to define mappings of user actions at different workflow states to transitions in the HumanWorkFlowActionState(should probably change this name) contentPart. This mapping will be used when the workflow hits a (proposed)human workflow step.

My feeling is that to implement a workflow like that, both Orchard workflows as well as the proposed permission management will be useful.
In Plone terms, you would need both Content Rules and the DCWorkflow state graph.

E.g in the Negotiation stage, you might want only the Sales and Legal groups to have the View permission on Proposal items.
But in the "Closed Lost" stage, you might want the Pre-sales and Consultant groups to also have the View permission on Proposal items.

To manage this in Orchard currently, if I understand correctly, you would need to write ad hoc checks in templates or in layer rules to decide whether to render something.

> > E.g. on a Blog post, the mapping is updated as the post moves through workflow states:
> 
> > Draft
> > Owner: Can view, Can edit, Can submit for review (-> Pending review)
> > Editor: no permission
> > Anonymous: no permission
> > Pending review
> > Owner: Can view, Can withdraw (-> Draft)
> > Editor: Can view, Can publish (-> Published), Can submit for revision (-> Draft), Can archive (-> Archived)
> > Anonymous: no permission
> > Published
> > Owner: Can view
> > Editor: Can view, Can submit for revision (-> Draft), Can archive (-> Archived)
> > Anonymous: Can view
> > Archived
> > Owner: Can view
> > Editor: Can view, Can submit for revision (-> Draft)
> > Anonymous: no permission
> 
> This example workflow is what we're trying to support.

Cool.

> I don't think a role-permisison map needs to be updated on each transition for each workflow instance. The rules could be defined once and evaluated at every step.

I there's still some divergence in what we're thinking about. 
Access is always evaluated in terms of permission. I.e. does this user have the View permission on this item.
In addition, you may have ad hoc checks, e.g. the user does have View permission but in the Liquid template you conditionally omit some detail depending on context. 

> I lean towards introducing these changes in one or more separate modules for now so they can be optionally turned on and minimize conflict with the core of OC.

For sure.

**Comment by sebastienros** [comment]
@jean can you try to find a recorded video of the User Actions feature in O1 that uses Workflows and is exactly what you are describing?

Does anyone know when/where we demoed it ?

**Comment by jean** [comment]
@sebastienros [Orchard 1 - User Task Workflow](https://www.youtube.com/watch?v=iioKQkXUwXA) which was [linked by @bakintunde](https://gitter.im/OrchardCMS/OrchardCore?at=5d5f5a8b142faa4fbd530d4c) demonstrates creating buttons for users to trigger workflow actions, but doesn't touch on state-based permissions.

Orchard implements an activity-based workflow system (which is basically visual scripting to automate actions in the CMS). (Which is awesome and works great.)
I'm describing a graph of item states, where each state defines a role/permission mapping (as in `OrchardCore.Roles/Admin/Index`) (and this graph is confusingly also called a workflow). 
The touchpoint between the activity-based workflow and the state-based workflow is that there should be ways to attempt to trigger a state transition (if the transition's guard conditions are satisfied). `UserTask` is one such a way.

Orchard workflows are about scripting business processes. 
DCWorkflow is about managing item security.  

Possibly the security management can be implemented by defining a `Set Permissions` task. I haven't thought more about the implications of this.

**Comment by jean** [comment]
@bakintunde 

> Ability to extend the user's profile with any of OC's fields so that we can have things like date of birth, address etc.

What would be nice is if there were a `UserPart` that one could add to any content type definition, to enable items to represent users that can log in. 

> Define a hierarchical structure of groups that can contain OC users or other groups | To model organizational/team structures. E.g. We might have multiple users with the administrator role in different groups who are responsible for managing the users in their own group.

I disagree that the goal is to _model organizational/team structures_. I think it is to _manage permissions_ in a scalable way. The group defines the set of users who get certain roles on certain content items in certain item states. I.e. in my understanding the group of people with Edit permission on certain sets of items doesn't correspond to a specific team in the org chart.

**Comment by willnationsdev** [comment]
Hello @bakintunde. I was curious how your work on this was coming along (currently at bakintunde/OrchardCore@8a209b47b61036b89ed035e0a319d79bd262962a). I can't quite tell based on the commit history, but development appears to have stalled. Are there any particular obstacles you encountered that this Issue could be updated with?

I have a team that may be interested in OrchardCore. If we end up choosing to use it, this feature will be an absolute necessity for our projects.
